src/util.h

```diff
@@ -300,58 +300,70 @@ class MaybeStackBuffer {
     return buf_[index];
   }
 
+  // Size of the buffer
   size_t length() const {
     return length_;
   }
 
-  // Call to make sure enough space for `storage` entries is available.
-  // There can only be 1 call to AllocateSufficientStorage or Invalidate
-  // per instance.
+  // Actual maximum capacity of the buffer with which SetLength() can be called
+  // without first calling AllocateSufficientStorage()
+  size_t capacity() const {
+    return IsAllocated() ? capacity_ : kStackStorageSize;
+  }
+
+  // Make sure enough space for `storage` entries is available.
+  // This method can be called multiple times throughout the lifetime of the
+  // buffer, but once this has been called Invalidate() cannot be used.
   void AllocateSufficientStorage(size_t storage) {
-    if (storage <= kStackStorageSize) {
-      buf_ = buf_st_;
-    } else {
-      buf_ = Malloc<T>(storage);
+    if (storage > capacity()) {
+      T* allocatedPtr = IsAllocated() ? buf_ : nullptr;
+      buf_ = Realloc<T>(allocatedPtr, storage);
+      capacity_ = storage;
```

Comment:
PR updated. Now contents up to `length_ - 1` are preserved if `buf_` was on the stack.